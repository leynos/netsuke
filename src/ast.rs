//! Netsuke manifest Abstract Syntax Tree structures.
//!
//! This module defines the data structures used to represent a parsed
//! `Netsukefile`. They mirror the YAML schema described in the design
//! document and are deserialised with `serde_yml`.
//!
//! The following example shows how to parse a minimal manifest string:
//!
//! ```rust
//! use netsuke::ast::NetsukeManifest;
//! use netsuke::ast::StringOrList;
//!
//! let yaml = "netsuke_version: \"1.0.0\"\ntargets:\n  - name: hello\n    recipe:\n      kind: command\n      command: \"echo hi\"";
//! let manifest: NetsukeManifest = serde_yml::from_str(yaml).expect("parse");
//! if let StringOrList::String(name) = &manifest.targets[0].name {
//!     assert_eq!(name, "hello");
//! }
//! ```

use semver::Version;
use serde::{Deserialize, de::Deserializer};

fn deserialize_actions<'de, D>(deserializer: D) -> Result<Vec<Target>, D::Error>
where
    D: Deserializer<'de>,
{
    let mut actions = Vec::<Target>::deserialize(deserializer)?;
    for action in &mut actions {
        action.phony = true;
    }
    Ok(actions)
}
use std::collections::HashMap;

/// Top-level manifest structure parsed from a `Netsukefile`.
///
/// Each field mirrors a key in the YAML manifest. Optional collections default
/// to empty to simplify deserialisation.
///
/// ```yaml
/// netsuke_version: "1.0.0"
/// actions: []
/// targets:
///   - name: hello
///     recipe:
///       kind: command
///       command: echo hi
/// ```
///
/// ```rust
/// use netsuke::ast::NetsukeManifest;
/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
/// let yaml = "netsuke_version: \"1.0.0\"\ntargets:\n  - name: hello\n    recipe:\n      kind: command\n      command: echo hi";
/// let manifest: NetsukeManifest = serde_yml::from_str(yaml)?;
/// assert_eq!(manifest.targets.len(), 1);
/// # Ok(()) }
/// ```
#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct NetsukeManifest {
    /// Semantic version of the manifest format.
    pub netsuke_version: Version,

    /// Global key/value pairs available to recipes.
    #[serde(default)]
    pub vars: HashMap<String, String>,

    /// Named rule templates that can be referenced by targets.
    #[serde(default)]
    pub rules: Vec<Rule>,

    /// Optional setup actions executed before normal targets. Each action is
    /// implicitly marked as `phony` during deserialisation.
    #[serde(default, deserialize_with = "deserialize_actions")]
    pub actions: Vec<Target>,

    /// Primary build targets.
    pub targets: Vec<Target>,

    /// Names of targets built when no command line target is supplied.
    #[serde(default)]
    pub defaults: Vec<String>,
}

/// A reusable command template.
///
/// A rule encapsulates a snippet of work that can be referenced by multiple
/// targets. It may define a command line, a script block, or delegate to another
/// named rule. Dependencies may be specified as either a single string or a
/// list of strings.
#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Rule {
    /// Unique identifier used by targets to reference this rule.
    pub name: String,
    /// The action executed when the rule is invoked.
    pub recipe: Recipe,
    /// Optional human-friendly summary.
    pub description: Option<String>,
    /// Prerequisites for the rule. Empty by default.
    #[serde(default)]
    pub deps: StringOrList,
}

/// Execution style for rules and targets.
///
/// The variant is selected using the `kind` field in the manifest. Each variant
/// corresponds to a different way of specifying how a command should run.
#[derive(Debug, Deserialize, Clone, PartialEq)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum Recipe {
    /// A single shell command.
    #[serde(alias = "command")]
    Command { command: String },
    /// An embedded multi-line script.
    #[serde(alias = "script")]
    Script { script: String },
    /// Invoke another named rule.
    #[serde(alias = "rule")]
    Rule { rule: StringOrList },
}

/// A single build target.
///
/// Targets describe the files produced by a rule and their dependencies.
/// `phony` targets are always considered out of date, while `always` targets are
/// regenerated even if their inputs are unchanged.
#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Target {
    /// Output file or files.
    pub name: StringOrList,
    /// How the target should be built.
    pub recipe: Recipe,

    /// Input files consumed by the recipe.
    #[serde(default)]
    pub sources: StringOrList,

    /// Normal prerequisites that must be built first.
    #[serde(default)]
    pub deps: StringOrList,

    /// Dependencies that do not cause a rebuild when changed.
    #[serde(default)]
    pub order_only_deps: StringOrList,

    /// Target-scoped variables available during command execution.
    #[serde(default)]
    pub vars: HashMap<String, String>,

    /// Declares that the target does not correspond to a real file.
    #[serde(default)]
    pub phony: bool,

    /// Force the recipe to run even if the outputs are up to date.
    #[serde(default)]
    pub always: bool,
}

/// A helper for fields that accept either a single string or a list of
/// strings.
///
/// It mirrors YAML syntax where a scalar or sequence is allowed. Empty values
/// deserialize to `StringOrList::Empty`.
///
/// ```yaml
/// # Scalar
/// name: hello
/// # Sequence
/// name:
///   - hello
///   - world
/// ```
#[derive(Debug, Deserialize, Default, Clone, PartialEq)]
#[serde(untagged)]
pub enum StringOrList {
    /// No value provided.
    #[default]
    Empty,
    /// A single string item.
    String(String),
    /// A list of string items.
    List(Vec<String>),
}
