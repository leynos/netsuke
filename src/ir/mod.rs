//! Intermediate Representation structures.
//!
//! This module defines the backend-agnostic build graph used by Netsuke after
//! validation. The IR mirrors the conceptual model of Ninja without embedding
//! any Ninja-specific syntax.
//!
//! # Examples
//!
//! ```
//! use netsuke::ir::{Action, BuildGraph, BuildEdge};
//! use netsuke::ast::Recipe;
//! use camino::Utf8PathBuf;
//!
//! let action = Action {
//!     recipe: Recipe::Command { command: "echo hi".into() },
//!     description: None,
//!     depfile: None,
//!     deps_format: None,
//!     pool: None,
//!     restat: false,
//! };
//! let mut graph = BuildGraph::default();
//! graph.actions.insert("a".into(), action);
//! graph.default_targets.push(Utf8PathBuf::from("hello"));
//! ```

mod cmd_interpolate;
mod cycle;
mod from_manifest;
mod graph;

pub use graph::{Action, BuildEdge, BuildGraph, IrGenError};
