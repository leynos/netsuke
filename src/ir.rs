//! Intermediate Representation structures.
//!
//! This module defines the backend-agnostic build graph used by Netsuke after
//! validation. The IR mirrors the conceptual model of Ninja without embedding
//! any Ninja-specific syntax.
//!
//! # Examples
//!
//! ```
//! use netsuke::ir::{Action, BuildGraph, BuildEdge};
//! use netsuke::ast::Recipe;
//! use std::path::PathBuf;
//!
//! let action = Action {
//!     recipe: Recipe::Command { command: "echo hi".into() },
//!     description: None,
//!     depfile: None,
//!     deps_format: None,
//!     pool: None,
//!     restat: false,
//! };
//! let mut graph = BuildGraph::default();
//! graph.actions.insert("a".into(), action);
//! graph.default_targets.push(PathBuf::from("hello"));
//! ```
//
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;

/// The complete, static build graph.
#[derive(Debug, Default, Clone)]
pub struct BuildGraph {
    /// All unique actions in the build keyed by a stable hash.
    pub actions: HashMap<String, Action>,
    /// All target files to be built keyed by output path.
    pub targets: HashMap<PathBuf, BuildEdge>,
    /// Targets built when no explicit target is requested.
    pub default_targets: Vec<PathBuf>,
}

/// A reusable command analogous to a Ninja rule.
use serde::Serialize;

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Action {
    pub recipe: Recipe,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub depfile: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub deps_format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pool: Option<String>,
    pub restat: bool,
}

/// A single build statement connecting inputs to outputs.
#[derive(Debug, Clone, PartialEq)]
pub struct BuildEdge {
    /// Identifier of the [`Action`] used for this edge.
    pub action_id: String,
    /// Explicit inputs that trigger a rebuild when changed.
    pub inputs: Vec<PathBuf>,
    /// Outputs explicitly generated by the command.
    pub explicit_outputs: Vec<PathBuf>,
    /// Outputs implicitly generated by the command (Ninja `|`).
    pub implicit_outputs: Vec<PathBuf>,
    /// Order-only dependencies that do not trigger rebuilds (Ninja `||`).
    pub order_only_deps: Vec<PathBuf>,
    /// Output does not correspond to a real file.
    pub phony: bool,
    /// Run the command on every invocation regardless of timestamps.
    pub always: bool,
}

use crate::ast::{NetsukeManifest, Recipe, Rule, StringOrList};
use shell_quote::{QuoteRefExt, Sh};
use thiserror::Error;

use crate::hasher::ActionHasher;

/// Errors produced during IR generation.
#[derive(Debug, Error)]
pub enum IrGenError {
    #[error("rule '{rule_name}' referenced by target '{target_name}' was not found")]
    RuleNotFound {
        target_name: String,
        rule_name: String,
    },

    #[error("multiple rules for target '{target_name}': {rules:?}")]
    MultipleRules {
        target_name: String,
        rules: Vec<String>,
    },

    #[error("No rules specified for target {target_name}")]
    EmptyRule { target_name: String },

    #[error("duplicate target outputs: {outputs:?}")]
    DuplicateOutput { outputs: Vec<String> },

    #[error("circular dependency detected: {cycle:?}")]
    CircularDependency { cycle: Vec<PathBuf> },

    #[error("failed to serialise action: {0}")]
    ActionSerialisation(#[from] serde_json::Error),

    #[error("command is not a valid shell command: {snippet}")]
    InvalidCommand { command: String, snippet: String },
}

impl BuildGraph {
    /// Transform a manifest into a [`BuildGraph`].
    ///
    /// # Errors
    ///
    /// Returns [`IrGenError`] when a referenced rule is missing, multiple rules
    /// are specified for a single target, or no rule is provided.
    pub fn from_manifest(manifest: &NetsukeManifest) -> Result<Self, IrGenError> {
        let mut graph = Self::default();
        let mut rule_map = HashMap::new();

        Self::process_rules(manifest, &mut rule_map);
        Self::process_targets(manifest, &mut graph.actions, &mut graph.targets, &rule_map)?;
        Self::process_defaults(manifest, &mut graph.default_targets);

        graph.detect_cycles()?;

        Ok(graph)
    }

    fn process_rules(manifest: &NetsukeManifest, rule_map: &mut HashMap<String, Arc<Rule>>) {
        for rule in &manifest.rules {
            rule_map.insert(rule.name.clone(), Arc::new(rule.clone()));
        }
    }

    fn process_targets(
        manifest: &NetsukeManifest,
        actions: &mut HashMap<String, Action>,
        targets: &mut HashMap<PathBuf, BuildEdge>,
        rule_map: &HashMap<String, Arc<Rule>>,
    ) -> Result<(), IrGenError> {
        for target in manifest.actions.iter().chain(&manifest.targets) {
            let outputs = to_paths(&target.name);
            let inputs = to_paths(&target.sources);
            let target_name = get_target_display_name(&outputs);
            let action_id = match &target.recipe {
                Recipe::Rule { rule } => {
                    // Rule recipe and description take precedence over any
                    // target-level fields. Future schema changes introducing
                    // explicit overrides should adjust this behaviour.
                    let tmpl = resolve_rule(rule, rule_map, &target_name)?;
                    register_action(
                        actions,
                        tmpl.recipe.clone(),
                        tmpl.description.clone(),
                        &inputs,
                        &outputs,
                    )?
                }
                Recipe::Command { .. } | Recipe::Script { .. } => {
                    register_action(actions, target.recipe.clone(), None, &inputs, &outputs)?
                }
            };

            let edge = BuildEdge {
                action_id,
                inputs: inputs.clone(),
                explicit_outputs: outputs.clone(),
                implicit_outputs: Vec::new(),
                order_only_deps: to_paths(&target.order_only_deps),
                phony: target.phony,
                always: target.always,
            };

            if let Some(dups) = find_duplicates(&outputs, targets) {
                return Err(IrGenError::DuplicateOutput { outputs: dups });
            }
            for out in outputs {
                targets.insert(out, edge.clone());
            }
        }
        Ok(())
    }

    fn process_defaults(manifest: &NetsukeManifest, defaults: &mut Vec<PathBuf>) {
        for name in &manifest.defaults {
            defaults.push(PathBuf::from(name));
        }
    }

    fn detect_cycles(&self) -> Result<(), IrGenError> {
        if let Some(cycle) = find_cycle(&self.targets) {
            return Err(IrGenError::CircularDependency { cycle });
        }
        Ok(())
    }
}

fn register_action(
    actions: &mut HashMap<String, Action>,
    recipe: Recipe,
    description: Option<String>,
    inputs: &[PathBuf],
    outputs: &[PathBuf],
) -> Result<String, IrGenError> {
    let recipe = match recipe {
        Recipe::Command { command } => {
            let interpolated = interpolate_command(&command, inputs, outputs)?;
            Recipe::Command {
                command: interpolated,
            }
        }
        other => other,
    };
    let action = Action {
        recipe,
        description,
        depfile: None,
        deps_format: None,
        pool: None,
        restat: false,
    };
    let hash = ActionHasher::hash(&action).map_err(IrGenError::ActionSerialisation)?;
    actions.entry(hash.clone()).or_insert(action);
    Ok(hash)
}

fn interpolate_command(
    template: &str,
    inputs: &[PathBuf],
    outputs: &[PathBuf],
) -> Result<String, IrGenError> {
    fn quote_paths(paths: &[PathBuf]) -> Vec<String> {
        paths
            .iter()
            .map(|p| {
                let raw = p.display().to_string();
                String::from_utf8(raw.quoted(Sh)).expect("quoted path is valid UTF-8")
            })
            .collect()
    }

    let ins = quote_paths(inputs).join(" ");
    let outs = quote_paths(outputs).join(" ");
    let interpolated = substitute(template, &ins, &outs);
    if shlex::split(&interpolated).is_none() {
        let snippet: String = interpolated.chars().take(160).collect();
        return Err(IrGenError::InvalidCommand {
            command: interpolated,
            snippet,
        });
    }
    Ok(interpolated)
}

/// Replace `$in`/`$out` tokens with pre-joined paths.
///
/// The scan is boundary-aware to avoid rewriting longer identifiers such as
/// `$input` or `$output_dir`.
fn substitute(template: &str, ins: &str, outs: &str) -> String {
    fn is_ident_byte(b: u8) -> bool {
        b.is_ascii_alphanumeric() || b == b'_'
    }

    fn is_boundary_safe(bytes: &[u8], i: usize, len: usize) -> bool {
        bytes
            .get(i.wrapping_sub(1))
            .is_none_or(|b| !is_ident_byte(*b))
            && bytes.get(i + len).is_none_or(|b| !is_ident_byte(*b))
    }

    fn matches_in_token(bytes: &[u8], i: usize) -> bool {
        bytes.get(i + 1) == Some(&b'i')
            && bytes.get(i + 2) == Some(&b'n')
            && is_boundary_safe(bytes, i, 3)
    }

    fn matches_out_token(bytes: &[u8], i: usize) -> bool {
        bytes.get(i + 1) == Some(&b'o')
            && bytes.get(i + 2) == Some(&b'u')
            && bytes.get(i + 3) == Some(&b't')
            && is_boundary_safe(bytes, i, 4)
    }

    fn match_macro<'a>(
        bytes: &[u8],
        i: usize,
        ins: &'a str,
        outs: &'a str,
    ) -> Option<(usize, &'a str)> {
        if matches_in_token(bytes, i) {
            return Some((3, ins));
        }
        if matches_out_token(bytes, i) {
            return Some((4, outs));
        }
        None
    }

    let bytes = template.as_bytes();
    let mut out = String::with_capacity(template.len() + ins.len() + outs.len());
    let mut i = 0;
    while let Some(&b) = bytes.get(i) {
        if b == b'$'
            && let Some((skip, repl)) = match_macro(bytes, i, ins, outs)
        {
            out.push_str(repl);
            i += skip;
            continue;
        }
        out.push(b as char);
        i += 1;
    }
    out
}

fn map_string_or_list<T, F>(sol: &StringOrList, f: F) -> Vec<T>
where
    F: Fn(&str) -> T,
{
    match sol {
        StringOrList::Empty => Vec::new(),
        StringOrList::String(s) => vec![f(s)],
        StringOrList::List(v) => v.iter().map(|s| f(s)).collect(),
    }
}

fn to_paths(sol: &StringOrList) -> Vec<PathBuf> {
    map_string_or_list(sol, |s| PathBuf::from(s))
}

fn to_string_vec(sol: &StringOrList) -> Vec<String> {
    map_string_or_list(sol, str::to_string)
}

fn extract_single(sol: &StringOrList) -> Option<&str> {
    match sol {
        StringOrList::String(s) => Some(s),
        StringOrList::List(v) if v.len() == 1 => v.first().map(String::as_str),
        _ => None,
    }
}

fn resolve_rule(
    rule: &StringOrList,
    rule_map: &HashMap<String, Arc<Rule>>,
    target_name: &str,
) -> Result<Arc<Rule>, IrGenError> {
    extract_single(rule).map_or_else(
        || {
            let mut rules = to_string_vec(rule);
            if rules.is_empty() {
                Err(IrGenError::EmptyRule {
                    target_name: target_name.to_string(),
                })
            } else {
                rules.sort();
                Err(IrGenError::MultipleRules {
                    target_name: target_name.to_string(),
                    rules,
                })
            }
        },
        |name| {
            rule_map
                .get(name)
                .cloned()
                .ok_or_else(|| IrGenError::RuleNotFound {
                    target_name: target_name.to_string(),
                    rule_name: name.to_string(),
                })
        },
    )
}

fn find_duplicates(
    outputs: &[PathBuf],
    targets: &HashMap<PathBuf, BuildEdge>,
) -> Option<Vec<String>> {
    let mut dups: Vec<_> = outputs
        .iter()
        .filter(|o| targets.contains_key(*o))
        .map(|o| o.display().to_string())
        .collect();
    if dups.is_empty() {
        None
    } else {
        dups.sort();
        Some(dups)
    }
}

fn get_target_display_name(paths: &[PathBuf]) -> String {
    paths
        .first()
        .map(|p| p.display().to_string())
        .unwrap_or_default()
}

#[derive(Clone, Copy)]
enum VisitState {
    Visiting,
    Visited,
}

fn should_visit_node<'a>(
    states: &'a mut HashMap<PathBuf, VisitState>,
    node: &'a PathBuf,
) -> Result<bool, &'a PathBuf> {
    match states.get(node) {
        Some(VisitState::Visited) => Ok(false),
        Some(VisitState::Visiting) => Err(node),
        None => {
            states.insert(node.clone(), VisitState::Visiting);
            Ok(true)
        }
    }
}

fn find_cycle(targets: &HashMap<PathBuf, BuildEdge>) -> Option<Vec<PathBuf>> {
    fn visit(
        targets: &HashMap<PathBuf, BuildEdge>,
        node: &PathBuf,
        stack: &mut Vec<PathBuf>,
        states: &mut HashMap<PathBuf, VisitState>,
    ) -> Option<Vec<PathBuf>> {
        match should_visit_node(states, node) {
            Ok(false) => return None,
            Err(path) => {
                if let Some(idx) = stack.iter().position(|n| n == path) {
                    let mut cycle = stack.get(idx..).expect("slice").to_vec();
                    cycle.push(path.clone());
                    return Some(canonicalize_cycle(cycle));
                }
                return Some(vec![path.clone(), path.clone()]);
            }
            Ok(true) => {}
        }

        stack.push(node.clone());

        if let Some(cycle) = targets
            .get(node)
            .and_then(|edge| visit_dependencies(targets, &edge.inputs, stack, states))
        {
            return Some(cycle);
        }

        stack.pop();
        states.insert(node.clone(), VisitState::Visited);
        None
    }

    fn visit_dependencies(
        targets: &HashMap<PathBuf, BuildEdge>,
        deps: &[PathBuf],
        stack: &mut Vec<PathBuf>,
        states: &mut HashMap<PathBuf, VisitState>,
    ) -> Option<Vec<PathBuf>> {
        for dep in deps {
            if !targets.contains_key(dep) {
                continue;
            }

            if let Some(cycle) = visit(targets, dep, stack, states) {
                return Some(cycle);
            }
        }
        None
    }

    let mut states = HashMap::new();
    let mut stack = Vec::new();

    for node in targets.keys() {
        // Skip nodes we've already processed to avoid redundant traversal.
        if states.contains_key(node) {
            continue;
        }
        if let Some(cycle) = visit(targets, node, &mut stack, &mut states) {
            return Some(cycle);
        }
    }
    None
}

fn canonicalize_cycle(mut cycle: Vec<PathBuf>) -> Vec<PathBuf> {
    if cycle.len() < 2 {
        return cycle;
    }
    let len = cycle.len() - 1;
    let start = cycle
        .iter()
        .take(len)
        .enumerate()
        .min_by(|(_, a), (_, b)| a.cmp(b))
        .map_or(0, |(idx, _)| idx);
    cycle.rotate_left(start);
    if let (Some(first), Some(slot)) = (cycle.first().cloned(), cycle.get_mut(len)) {
        slot.clone_from(&first);
    }
    cycle
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn find_cycle_identifies_cycle() {
        let mut targets = HashMap::new();
        let edge_a = BuildEdge {
            action_id: "id".into(),
            inputs: vec![PathBuf::from("b")],
            explicit_outputs: vec![PathBuf::from("a")],
            implicit_outputs: Vec::new(),
            order_only_deps: Vec::new(),
            phony: false,
            always: false,
        };
        let edge_b = BuildEdge {
            action_id: "id".into(),
            inputs: vec![PathBuf::from("a")],
            explicit_outputs: vec![PathBuf::from("b")],
            implicit_outputs: Vec::new(),
            order_only_deps: Vec::new(),
            phony: false,
            always: false,
        };
        targets.insert(PathBuf::from("a"), edge_a);
        targets.insert(PathBuf::from("b"), edge_b);

        let cycle = find_cycle(&targets).expect("cycle");
        let option_a = vec![PathBuf::from("a"), PathBuf::from("b"), PathBuf::from("a")];
        let option_b = vec![PathBuf::from("b"), PathBuf::from("a"), PathBuf::from("b")];
        assert!(cycle == option_a || cycle == option_b);
    }
}
