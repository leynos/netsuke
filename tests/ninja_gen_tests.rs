//! Unit tests for Ninja file generation.
//!
//! Snapshot tests use `insta` to ensure the emitted manifest remains stable.
//! Tests cover various scenarios including phony targets, standard builds with
//! multiple inputs and outputs, complex dependency relationships, and edge
//! cases like empty build graphs.

use insta::{Settings, assert_snapshot};
use netsuke::ast::Recipe;
use netsuke::ir::{Action, BuildEdge, BuildGraph};
use netsuke::ninja_gen::generate;
use rstest::rstest;
use std::{fs, path::PathBuf, process::Command};
use tempfile::tempdir;

fn skip_if_ninja_unavailable() -> bool {
    match Command::new("ninja").arg("--version").output() {
        Err(_) => {
            eprintln!("skipping test: ninja not found in PATH");
            true
        }
        Ok(output) if !output.status.success() => {
            eprintln!("skipping test: ninja --version failed");
            true
        }
        Ok(_) => false,
    }
}

#[rstest]
fn generate_phony() {
    let action = Action {
        recipe: Recipe::Command {
            command: "true".into(),
        },
        description: None,
        depfile: None,
        deps_format: None,
        pool: None,
        restat: false,
    };
    let edge = BuildEdge {
        action_id: "a".into(),
        inputs: vec![PathBuf::from("in")],
        explicit_outputs: vec![PathBuf::from("out")],
        implicit_outputs: Vec::new(),
        order_only_deps: Vec::new(),
        phony: true,
        always: false,
    };
    let mut graph = BuildGraph::default();
    graph.actions.insert("a".into(), action);
    graph.targets.insert(PathBuf::from("out"), edge);

    let ninja = generate(&graph);
    let expected = concat!(
        "rule a\n",
        "  command = true\n\n",
        "build out: phony in\n\n",
    );
    assert_eq!(ninja, expected);
}

#[rstest]
fn generate_standard_build() {
    let action = Action {
        recipe: Recipe::Command {
            command: "cc -c $in -o $out".into(),
        },
        description: None,
        depfile: None,
        deps_format: None,
        pool: None,
        restat: false,
    };
    let edge = BuildEdge {
        action_id: "compile".into(),
        inputs: vec![PathBuf::from("a.c"), PathBuf::from("b.c")],
        explicit_outputs: vec![PathBuf::from("ab.o")],
        implicit_outputs: Vec::new(),
        order_only_deps: Vec::new(),
        phony: false,
        always: false,
    };
    let mut graph = BuildGraph::default();
    graph.actions.insert("compile".into(), action);
    graph.targets.insert(PathBuf::from("ab.o"), edge);

    let ninja = generate(&graph);
    let expected = concat!(
        "rule compile\n",
        "  command = cc -c $in -o $out\n\n",
        "build ab.o: compile a.c b.c\n\n",
    );
    assert_eq!(ninja, expected);
}

#[rstest]
fn generate_complex_dependencies() {
    let action = Action {
        recipe: Recipe::Command {
            command: "true".into(),
        },
        description: None,
        depfile: None,
        deps_format: None,
        pool: None,
        restat: false,
    };
    let edge = BuildEdge {
        action_id: "b".into(),
        inputs: vec![PathBuf::from("in")],
        explicit_outputs: vec![PathBuf::from("out"), PathBuf::from("log")],
        implicit_outputs: vec![PathBuf::from("out.d")],
        order_only_deps: vec![PathBuf::from("stamp")],
        phony: false,
        always: false,
    };
    let mut graph = BuildGraph::default();
    graph.actions.insert("b".into(), action);
    graph.targets.insert(PathBuf::from("out"), edge);

    let ninja = generate(&graph);
    let expected = concat!(
        "rule b\n",
        "  command = true\n\n",
        "build out log | out.d: b in || stamp\n\n",
    );
    assert_eq!(ninja, expected);
}

#[rstest]
fn generate_empty_graph() {
    let graph = BuildGraph::default();
    let ninja = generate(&graph);
    assert!(ninja.is_empty());
}

#[rstest]
fn generate_multiline_script_snapshot() {
    let mut graph = BuildGraph::default();
    graph.actions.insert(
        "script".into(),
        Action {
            recipe: Recipe::Script {
                script: "echo one\necho two".into(),
            },
            description: None,
            depfile: None,
            deps_format: None,
            pool: None,
            restat: false,
        },
    );
    graph.targets.insert(
        PathBuf::from("out"),
        BuildEdge {
            action_id: "script".into(),
            inputs: Vec::new(),
            explicit_outputs: vec![PathBuf::from("out")],
            implicit_outputs: Vec::new(),
            order_only_deps: Vec::new(),
            phony: false,
            always: false,
        },
    );
    graph.default_targets.push(PathBuf::from("out"));

    let ninja = generate(&graph);
    let mut settings = Settings::new();
    settings.set_snapshot_path(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/tests/snapshots/ninja",
    ));
    settings.bind(|| {
        assert_snapshot!("multiline_script_ninja", &ninja);
    });
    let lines: Vec<&str> = ninja.lines().collect();
    let command_line = lines.get(1).expect("command line");
    assert!(
        command_line.contains("\\n"),
        "script newline should be encoded",
    );
}

/// Ensure a multi-line script produces a Ninja manifest that Ninja accepts.
#[rstest]
#[ignore = "requires Ninja"]
fn integration_multiline_script_valid() {
    if skip_if_ninja_unavailable() {
        return;
    }

    let mut graph = BuildGraph::default();
    graph.actions.insert(
        "script".into(),
        Action {
            recipe: Recipe::Script {
                script: "echo one\necho two".into(),
            },
            description: None,
            depfile: None,
            deps_format: None,
            pool: None,
            restat: false,
        },
    );
    graph.targets.insert(
        PathBuf::from("out"),
        BuildEdge {
            action_id: "script".into(),
            inputs: Vec::new(),
            explicit_outputs: vec![PathBuf::from("out")],
            implicit_outputs: Vec::new(),
            order_only_deps: Vec::new(),
            phony: false,
            always: false,
        },
    );
    graph.default_targets.push(PathBuf::from("out"));

    let ninja = generate(&graph);
    let dir = tempdir().expect("temp dir");
    fs::write(dir.path().join("build.ninja"), &ninja).expect("write ninja");
    let status = Command::new("ninja")
        .arg("-n")
        .current_dir(dir.path())
        .status()
        .expect("run ninja");
    assert!(status.success());
}

/// Test that scripts containing percent signs execute correctly.
#[rstest]
#[ignore = "requires Ninja"]
fn generate_script_with_percent() {
    if skip_if_ninja_unavailable() {
        return;
    }

    let action = Action {
        recipe: Recipe::Script {
            script: "echo 100% > out".into(),
        },
        description: None,
        depfile: None,
        deps_format: None,
        pool: None,
        restat: false,
    };
    let edge = BuildEdge {
        action_id: "percent".into(),
        inputs: Vec::new(),
        explicit_outputs: vec![PathBuf::from("out")],
        implicit_outputs: Vec::new(),
        order_only_deps: Vec::new(),
        phony: false,
        always: false,
    };
    let mut graph = BuildGraph::default();
    graph.actions.insert("percent".into(), action);
    graph.targets.insert(PathBuf::from("out"), edge);
    graph.default_targets.push(PathBuf::from("out"));

    let ninja = generate(&graph);
    let dir = tempdir().expect("temp dir");
    fs::write(dir.path().join("build.ninja"), &ninja).expect("write ninja");
    let status = Command::new("ninja")
        .arg("out")
        .current_dir(dir.path())
        .status()
        .expect("run ninja");
    assert!(status.success());
    let content = fs::read_to_string(dir.path().join("out")).expect("read out");
    assert_eq!(content.trim(), "100%");
}
