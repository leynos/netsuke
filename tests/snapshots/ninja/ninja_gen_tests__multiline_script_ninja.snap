---
source: tests/ninja_gen_tests.rs
assertion_line: 176
expression: "&ninja"
---
rule script
  command = /bin/sh -e -c "printf %b 'echo one\necho two' | /bin/sh -e"

build out: script

default out
